lib/ReplaceOpenCLBuiltinPass.cpp

```diff
@@ -678,6 +681,161 @@ Type *ReplaceOpenCLBuiltinPass::GetPairStruct(Type *type) {
   return new_struct;
 }
 
+Value *ReplaceOpenCLBuiltinPass::InsertOpMulExtended(Instruction *InsertPoint,
+                                                     Value *a, Value *b,
+                                                     bool IsSigned) {
+
+  Type *Ty = a->getType();
+  Type *RetTy = GetPairStruct(a->getType());
+  assert(Ty == b->getType());
+
+  if (!Option::HackMulExtended()) {
+    spv::Op opcode = IsSigned ? spv::OpSMulExtended : spv::OpUMulExtended;
+
+    return clspv::InsertSPIRVOp(InsertPoint, opcode, {Attribute::ReadNone},
+                                RetTy, {a, b});
+  }
+
+  unsigned int ScalarSizeInBits = Ty->getScalarSizeInBits();
+  bool IsVector = Ty->isVectorTy();
+
+  IRBuilder<> Builder(InsertPoint);
+
+  if (ScalarSizeInBits <= 32) {
```

Comment:
Doing this for 32-bit integers means adding the Int64 capability requirement. That seems unfortunate for a workaround.

Reply:
I have modified it so that 32x32 in 64 is done as 64x64 in 128. But the 64x64 in 128 is still using Int64 as it makes sense to require it in that case.